torch/utils/pytree.py

```diff
@@ -0,0 +1,419 @@
+"""
+Contains utility functions for working with nested python data structures.
+
+A *pytree* is Python nested data structure. It is a tree in the sense that
+nodes are Python collections (e.g., list, tuple, dict) and the leaves are
+Python values. Furthermore, a pytree should not contain reference cycles.
+
+pytrees are useful for working with nested collections of Tensors. For example,
+one can use `tree_map` to map a function over all Tensors inside some nested
+collection of Tensors and `tree_leaves` to get a flat list of all Tensors
+inside some nested collection. pytrees are helpful for implementing nested
+collection support for PyTorch APIs.
+"""
+
+import functools
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    List,
+    Optional,
+    overload,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+import optree
+from optree import PyTreeSpec  # direct import for type annotations
+
+
+__all__ = [
+    "PyTree",
+    "PyTreeSpec",
+    "register_pytree_node",
+    "tree_flatten",
+    "tree_unflatten",
+    "tree_leaves",
+    "tree_structure",
+    "tree_map",
+    "tree_map_",
+    "tree_map_only",
+    "tree_map_only_",
+    "tree_all",
+    "tree_any",
+    "tree_all_only",
+    "tree_any_only",
+    "broadcast_prefix",
+    "_broadcast_to_and_flatten",
+]
+
+
+T = TypeVar("T")
+S = TypeVar("S")
+R = TypeVar("R")
+
+
+Context = Any
+PyTree = Any
+TreeSpec = PyTreeSpec
+FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
```

Comment:
```suggestion
    FlattenFunc = Callable[[PyTree], Tuple[List, Optional[Context]]]
```
based on your other file? These mypy type notations aren't consistent with usage in torch.fx